File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.py
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        left = 0
        ans = 0
        odd_count = 0
        cur_sub_count = 0
        
        for right in range(len(nums)):
            # If the current element is odd, increase the odd_count and reset cur_sub_count
            if nums[right] % 2 == 1:
                odd_count += 1
                cur_sub_count = 0
            
            # When we have exactly k odd numbers in the window
            while odd_count == k:
                # If the left element is odd, decrement odd_count
                if nums[left] % 2 == 1:
                    odd_count -= 1
                # Increase the count of valid subarrays ending at right
                cur_sub_count += 1
                # Move the left pointer to shrink the window
                left += 1
            
            # Add the number of valid subarrays ending at right to the answer
            ans += cur_sub_count
        
        return ans

File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        return helper(nums,k) - helper(nums,k-1);
    }


    public int helper(int[] nums,int k){
         int left = 0;
         int count = 0;
         int result = 0;
         for(int right = 0;right<nums.length;right++){
            if(nums[right]%2==1){
                count++;
            }

            while(count > k ){
                if( nums[left]%2 == 1) count--;
                left++;
            }

            if(count <= k){
                result += right - left +1; 
            }
         }
         return result;
    }
}